src/fpm.f90

```diff
@@ -203,6 +192,138 @@ subroutine build_model(model, settings, package, error)
     if (duplicates_found) then
         call fpm_stop(1,'*build_model*:Error: One or more duplicate module names found.')
     end if
+
+    ! Compiler flags logic
+    if(settings%profile.eq.'')then
+        if (trim(settings%flag).eq.'') then
+            profile = 'debug'
+        end if
+    else
+        profile = settings%profile
+    endif
+
+    ! Choose profile for each package
+    if (allocated(profile)) then
+        do i=1,size(model%packages)
+            model%packages(i)%chosen_profile = look_for_profile(i)
+        end do
+    end if
+
+    model%archiver = get_archiver()
+    call get_default_c_compiler(model%fortran_compiler, model%c_compiler)
+    model%c_compiler = get_env('FPM_C_COMPILER',model%c_compiler)
+
+    if (is_unknown_compiler(model%fortran_compiler)) then
+        write(*, '(*(a:,1x))') &
+            "<WARN>", "Unknown compiler", model%fortran_compiler, "requested!", &
+            "Defaults for this compiler might be incorrect"
+    end if
+
+    ! Choose profiles flags or file specific flags
+    do j=1,size(model%packages)
+        associate(package=>model%packages(j), sources=>model%packages(j)%sources, profile=>model%packages(j)%chosen_profile)
+            do i=1,size(sources)
+                select case (sources(i)%unit_type)
+                case (FPM_UNIT_MODULE,FPM_UNIT_SUBMODULE,FPM_UNIT_SUBPROGRAM,FPM_UNIT_CSOURCE)
+                    file_scope_flag = get_file_scope_flags(sources(i), profile)
+                    if (sources(i)%unit_type.eq.FPM_UNIT_CSOURCE) then
+                        if (file_scope_flag.eq."") then
+                            sources(i)%flags=model%cmd_compile_flags//" "//profile%c_flags
+                        else
+                            sources(i)%flags=model%cmd_compile_flags//" "//file_scope_flag
+                        end if
+                    else
+                        if (file_scope_flag.eq."") then
+                            sources(i)%flags=model%cmd_compile_flags//" "//profile%flags
+                        else
+                            sources(i)%flags=model%cmd_compile_flags//" "//file_scope_flag
+                        end if
+                    end if
+                case (FPM_UNIT_PROGRAM)
+                    file_scope_flag = get_file_scope_flags(sources(i), profile)
+                    if (file_scope_flag.eq."") then
+                        sources(i)%flags=model%cmd_compile_flags//" "//profile%flags
+                    else
+                        sources(i)%flags=model%cmd_compile_flags//" "//file_scope_flag
+                    end if
+                    sources(i)%link_time_flags=profile%link_time_flags
+                end select
```

Comment:
There's a lot of duplication here which makes maintainability harder - I think this can be simplified, please see my suggestion below and check the logic

```suggestion
                case (FPM_UNIT_MODULE,FPM_UNIT_SUBMODULE,FPM_UNIT_SUBPROGRAM,FPM_UNIT_CSOURCE,FPM_UNIT_PROGRAM)
                    file_scope_flag = get_file_scope_flags(sources(i), profile)
                    if (file_scope_flag.eq."") then
                        if (sources(i)%unit_type.eq.FPM_UNIT_CSOURCE) then
                            sources(i)%flags=model%cmd_compile_flags//" "//profile%c_flags
                        else
                            sources(i)%flags=model%cmd_compile_flags//" "//profile%flags
                        end if
                    else
                        sources(i)%flags=model%cmd_compile_flags//" "//file_scope_flag
                    end if
                    
                    if (sources(i)%unit_type == FPM_UNIT_PROGRAM)
                        sources(i)%link_time_flags=profile%link_time_flags
                    end if
                end select
```